Doc tweaks for paper resubmission (#374)
vanvalenlab/kiosk-console: docs/source/DEVELOPER.rst

5. any Persistent Disks associated with your cluster

While we hope this list is comprehensive, there could be some lingering resources used by Google Cloud and not deleted automatically that we're not aware of.

Benchmarking Figures from DeepCell Kiosk Paper
---------------------------------------------------------

> The DeepCell Kiosk paper (Bannon et al., 2020) presents cost and runtime benchmarks for running a generic segmentation pipeline on datasets of given sizes inside the DeepCell Kiosk with different sets of constraints (Fig. 1b).

Reviewer: I think this should be "Benchmarking the cluster" and then later on how one can use the output files to recreate the graphs.

> The DeepCell Kiosk comes with a utility for benchmarking the scalability and performance of a deep learning workflow. To run your own benchmarking, please read below.

Author: I did this. But I realize now that I forgot to add in the figure recreation section at the end. I'll push that in a few minutes.

Author: Pushed.

Here, we provide any interested parties with a way to recreate these figures.

> Recreating Figures

Author: Point well taken. I tried chopping out words where possible.

> Recreating Figures

Author: Well, and then I ended up putting a lot more in. Someone stop me, please.

> 1. If you don't already have a cloud storage bucket devoted exclusively to DeepCell Kiosk benchmarking runs, you should create one now. It's fine to reuse this bucket for multiple benchmarking runs, but don't use it as the storage bucket for normal DeepCell Kiosk cluster operations.

Reviewer: Why shouldn't they use the bucket for normal operations?

Author: If someone uploads something to the bucket during a benchmarking run, won't that mess things up? How do you think we should express that warning?

Reviewer: It is an extra job that is being processed, but the benchmarking job will still be run.

Maybe just a warning that submitting extra jobs to a cluster during benchmarking will affect benchmarking performance.

^^^^^^^^^^^^^^^^^^

> 1. First, checkout the `benchmarks` branch of the `kiosk-console` repository. This branch is almost identical to the version 1.2.0 release of the `kiosk-console` repo, except for minor configuration differences, which the user can view by executing `git diff master..benchmarks` from a command prompt inside a local copy of the repository. There are three very important variable settings worth noting, each with a corresponding setup action for the user to perform before benchmarking:

Author: I simplified this section greatly, and fixed the formatting.

@willgraf
Yeah, we probably should just protect `benchmarks`. I took out the `git diff` line, but we can put a direct link to the diffs back in if you want.

> 1. First, checkout the `benchmarks` branch of the `kiosk-console` repository. This branch is almost identical to the version 1.2.0 release of the `kiosk-console` repo, except for minor configuration differences, which the user can view by executing `git diff master..benchmarks` from a command prompt inside a local copy of the repository. There are three very important variable settings worth noting, each with a corresponding setup action for the user to perform before benchmarking:

Reviewer: Yeah I do like the idea of having a link to the changes we made for our own benchmarking runs. I think that's a perfect way to explicitly show our own configuration.

> 2. Then, checkout the ``benchmarks`` branch of the ``kiosk-console`` repository. This branch is almost identical to the version 1.2.0 release of the ``kiosk-console`` repo, except for minor configuration differences. There are three variable settings worth noting in the benchmarking pod's YAML file, ``conf/helmfile.d/0410.benchmarking.yaml``, each with a corresponding setup action for the user to perform before benchmarking:

Reviewer: I'm not sure if we should recommend a `git checkout` but maybe just use it as a reference for our own configuration?

I mean to say that the `benchmarks` branch is the configuration we used, but they can just use it as a reference rather than needing to start from that branch. None of those settings is _required_ to be changed, but they just show what the settings we used are.

Author: Oh, so you want to tell people they can run benchamarking off master? Can we do that in the figure recreation section, too?

Reviewer: Yeah exactly, the benchmarking pod will work fine from `master`. The `benchmarks` branch just shows the configuration settings we used and sets them as default choices during the cluster config menu.

For figure recreation, I would just point them to the benchmarking branch as an example of all the required changes. Maybe still walk them through each change for explicit-ness.

>     - In the `Dockerfile`, the `CLOUDSDK_BUCKET` environmental variable is set to `kiosk-benchmarking`, which is the Google Cloud bucket that the Van Valen Lab uses to store its benchmarking models. To recreate our benchmarking data, please change this value to the name of a bucket that you control.

Reviewer 2: I'm a bit lost on the direction for both this point and the following. Is there an easier way to do/explain this?

Reviewer: Yeah they do not need to change the environmental variable, they are prompted to enter their own bucket during the kiosk configuration.

Author: Fixed.

>     - In the benchmarking pod's YAML file, `conf/helmfile.d/0410.benchmarking.yaml`, the `MODEL` environmental variable has been set to the exact model name and version we used in our benchmarking, `NuclearSegmentation:2`. The benchmarking pod will look for the model files corresponding to this name and version in the `models/` folder of your bucket. To recreate our benchmarking data, please copy all the files form the `models/NuclearSegmentation` folder in the `kiosk-benchmarking` bucket to the `models/NuclearSegmentation` folder in your bucket.

Reviewer: Technically the benchmarking pod doesn't look anywhere, it just creates a job with that model field defined, `tf-serving` is what looks for the model.

I think we can tighten pu the copy from/to line, somehow:

> Please copy the `models/NuclearSegmentation` folder from the `kiosk-benchmarks` bucket to your own bucket.

Typo `form` -> `from`.

Author: The reference to the "benchmarking pod" has been changed to the "benchmarking cluster". The other changes were incorporated verbatim.

>     - Also in the benchmarking pod's YAML file is an environmental variable, `FILE`, which is set to `zip100.zip`. The benchmarking pod will look for a file by this name in the bucket you specified above at the location `uploads/[file_name]`. We have a file named `zip100.zip` located in our `kiosk-bencharking` bucket at `sample_data/zip100.zip`. This file consists of 100 microscopy images, which we used as the basis for all our benchmarking runs. To recreate our benchmarking data, please copy `sample_data/zip100.zip` from the `kiosk-benchmarking` bucket to `uploads/zip100.zip` in your bucket.

Reviewer: Again, the benchmarking pod itself does not look to see it is there. Instead, it just creates a job assuming that the file exists. The redis consumer is the pod that actually checks for the file and downloads it.

Author: I think I removed this reference.

>     - A final variable in the benchmarking pod's YAML file is ``COUNT``, which determines how many times the ``zip100.zip`` file will be submitted to the cluster for processing. This controls the total number of images the cluster will process over the course of the benchmarking run. For example, if you want to do a 10,000-image benchmarking run, you would set ``COUNT`` TO 100, since 100*100 = 10,000. Benchmarking data was presented in the paper for 10,000-image, 100,000-image, and 1,000,000-image runs. Accordingly, depending on which subset of the paper's benchmarking data the user wishes to recreate, ``COUNT`` should be set to either 100, 1,000, or 10,000.

Reviewer: The intro section already says there are three variables in the benchmarking helmfile that need to be changed, we don't really need to explain in each item that the varaible is in the benchmarking helmfile. It might be simpler to just reformat as a bullet list:

> There are three variable settings worth noting in the benchmarking pod's YAML file, ``conf/helmfile.d/0410.benchmarking.yaml``, each with a corresponding setup action for the user to perform before benchmarking:
> - `MODEL` : the model name and version of the model in the bucket, e.g. `NuclearSegmentation:2`.
> - `FILE`: The datafile that will be processed by a deep learning job
> - `COUNT`: The number of jobs to create

Author: Yeah, the number of references to the YAML file can definitely be cut down. You think I should remove all the explanation on choosing values and copying files?

Reviewer: I think just tightening up the explanations will work. Maybe a template like:

> - `COUNT`: The number of jobs used for benchmarking. Values of 100, 1000,  10000 were used in the paper.

>     - A final variable in the benchamrking pod's YAML file is `COUNT`, which determines how many times the `zip100.zip` file will be submitted to the cluster for processing. This controls the total number of images the cluster will process over the course of the benchmarking run. For example, if you want to do a 10,000-image benchmarking run, you would set `COUNT` TO 100, since 100*100 = 10,000. Benchmarking data was presented in the paper for 10,000-image, 100,000-image, and 1,000,000-image runs. Accordingly, depending on which subset of our benchmarking data the user wishes to recreate, `COUNT` should be set to either 100, 1,000, or 10,000.

Reviewer 2: I don't feel the last 2 sentences are necessary here.

Author: I definitely want to cut down the text length, but I'm also worried about leaving ambiguity when instructing people on how to recreate the benchmarking figures.

Reviewer 2: It's not so much ambiguity as it is brevity and conciseness. We will provide enough information to recreate our benchmarks. If they are unable to after that, then it is likely that will receive an email stating so.

Reviewer: I mentioned this elsewhere, but I think it makes a lot of sense to separate this into 2 sections:
- How to run the benchmarking pod
- How to duplicate our data.

> 2. Deploy a DeepCell Kiosk as you normally would. While navigating the cluster configuration menu, keep in mind that the option for Maximum Number of GPUs is relevant to the benchmarking process. In the paper, benchmarking data was presented for clusters with maxima of 1, 4, and 8 GPUs. Choose the appropriate maximum for the benchmarking dataset you would like to recreate.

Reviewer 2: Is this necessary?

Author: Heavily reorganized.

>     - The Maximum Number of GPUs is relevant to the benchmarking process. In the paper, benchmarking data was presented for clusters with maxima of 1, 4, and 8 GPUs. Choose the appropriate maximum for the benchmarking dataset you would like to recreate.

Reviewer: I think this doc should have less a focus on our own benchmarking run, and more of a focus on how they can use the utility for their own benchmarking runs (which may or may not include replicating our own).

Author: Ok, in all other instances, I included an explicit statement that the user could choose whatever value they desired in addition to telling them what settings they'd need to pick to recreate the paper figure. What do you think about that approach?

> 4. Once the cluster has finished deploying, drop to the ``Shell`` via the DeepCell Kiosk main menu and execute ``Start the benchmarking pod by executing the following command in the shell: kubectl scale deployment benchmarking --replicas=1.`` to deploy the benchmarking pod and begin the benchmarking process.

Reviewer: ```suggestion
4. Once the cluster has finished deploying, drop to the ``Shell`` via the DeepCell Kiosk main menu and start the benchmarking pod by executing: ``kubectl scale deployment benchmarking --replicas=1.``.
```

Author: Yeah, sorry, this should be fixed in the most recent commit.

> 3. Once the cluster has finished deploying, drop to the shell via the Kiosk main menu and change the `values.replicas` field in `conf/helmfile.d/0410.benchmarking.yaml` from `0` to `1`. This tells the DeepCell Kiosk to deploy one benchmarking pod and, in the absence of any other input to the Kiosk (i.e., image submission via the command line, web interface, or ImageJ plugin), it will execute a full benchmarking run. Now redeploy the `benchmarking` deployment by executing `helmfile -l name=benchmarking sync`.

Author: I wasn't even aware of that command. Great improvement.

> 5. Once `becnhmarking` has redeployed, the benchmarking process will begin. Benchmarking jobs can take a day or more, depending on the conditions chosen. To monitor the status of your benchmarking job, drop to the shell within the DeepCell Kiosk and execute the command `kubectl get logs benchmarking -- tail 100`. This will show you the most recent log output form the `benchmarking` pod. When benchmarking has finished, the final line in the log should be `Uploaded [FILEPATH] to [BUCKET] in [SECONDS] seconds.`, where `[FILEPATH]` is the location in `[BUCKET]`  where the benchmarking data has been saved.

Author: Incorporated.

> 6. Now that data has been generated for your benchmarking run and saved in your bucket, you can analyze it. While we have a repository that generates the exact figures used in the paper (see the `version2` branch of the repository stored at github.com/vanvalenlab/publication-figures), that repository expects output files from all run conditions. Alternatively, if you just want to analyze a single run's data, you can download the output file from your bucket and look at two top-level fields in this large JSON file:

Reviewer 2: This feels like way too much information, some of which could be moved to the figures repo. We should update master there so its not branch specific. I also am not a big fan of the second bullet point - we should fix the underestimate and point them directly to function - something that I think would best be done in the figure repo.

Author: I got a very weird error when trying to `git clone` the `publication-figures` repo. (Something about hte pull working but the checkout failing?) Otherwise, I would have made a PR for this merge. Can one of you confirm that pulling on that repo works for you?

Reviewer: I just cloned the repository without an issue. I've also [submitted a PR](https://github.com/vanvalenlab/publication-figures/pull/1) for you to review and merge.

    - The exact running time of the benchmarking procedure is given in seconds as the value of the `time_elapsed` field.

>     - A slight underestimate of the total costs of the benchmarking run can be found as the value to the `total_node_and_networking_costs` field. (This is a slight underestimate because it came to our attention late in the benchmarking process that we, in fact, weren't including all possible costs in the this value. To see the formula we used to adjust the total costs upwards, please see the function `extra_network_costs` in the file `https://github.com/vanvalenlab/publication-figures/blob/version2/figure_generation/data_extractor.py`.)

Author: Absolutely.

5. any Persistent Disks associated with your cluster

While we hope this list is comprehensive, there could be some lingering resources used by Google Cloud and not deleted automatically that we're not aware of.

Benchmarking the DeepCell Kiosk
---------------------------------------------------------

> The DeepCell Kiosk comes with a utility for benchmarking the scalability and performance of a deep learning workflow. To reproduced the cost and timing benchmarks reported in the DeepCell Kiosk paper, please refer to our `figure creation repository <https://github.com/vanvalenlab/kiosk-console/pull/github.com/vanvalenlab/publication-figures>`_. To run your own benchmarking, please read below.

Reviewer: ```suggestion
The DeepCell Kiosk comes with a utility for benchmarking the scalability and performance of a deep learning workflow. To reproduce the cost and timing benchmarks reported in the DeepCell Kiosk paper, please refer to our `figure creation repository <https://github.com/vanvalenlab/kiosk-console/pull/github.com/vanvalenlab/publication-figures>`_. To run your own benchmarking, please read below.
```

> 1. If you don't already have a cloud storage bucke for use with the DeepCell Kiosk, you should create one now. It's fine to reuse this bucket across multiple DeepCell Kiosk clusters.

Reviewer: ```suggestion
1. If you don't already have a cloud storage bucket for use with the DeepCell Kiosk, you should create one now. It's fine to reuse this bucket across multiple DeepCell Kiosk clusters.
```

> 1. If you don't already have a cloud storage bucket devoted exclusively to DeepCell Kiosk benchmarking runs, you should create one now. It's fine to reuse this bucket for multiple benchmarking runs, but don't use it as a storage bucket for normal DeepCell Kiosk cluster operations.

Reviewer: Correcting my previous comment which I can't readily find, there is no reason to not share the bucket. Uploads to the bucket do not trigger jobs; jobs are triggered by hitting the cluster's API. There is no risk for sharing the bucket across multiple clusters.

> 2. Checkout the ``benchmarks`` branch of the ``kiosk-console`` repository. This branch is almost identical to the version 1.2.0 release of the ``kiosk-console`` repo, except for minor configuration differences. There are three variables in the benchmarking pod's YAML file, ``conf/helmfile.d/0410.benchmarking.yaml``, that may need to be customized before benchmarking:

Reviewer: Reading it, the `benchmarks` branch reference doesn't seem useful here. Maybe we include that in the reproducibility section, and cut it from here?

```suggestion
2. There are three variables in the benchmarking pod's YAML file, ``conf/helmfile.d/0410.benchmarking.yaml``, that may need to be customized before benchmarking:
```

Author: What do you mean by the reproducibility section exactly?

>     - ``MODEL`` is the model name and version we used in benchmarking. The model you choose should be present in the ``models/`` folder of your benchmarking bucket. See the `Van Valen Lab's benchmarking bucket <https://console.cloud.google.com/storage/browser/kiosk-benchmarking>`_ for an example.

Reviewer: ```suggestion
    - ``MODEL`` is the model name and version that will be used in benchmarking. The model you choose should be present in the ``models/`` folder of your benchmarking bucket. See the `Van Valen Lab's benchmarking bucket <https://console.cloud.google.com/storage/browser/kiosk-benchmarking>`_ for an example.
```

>     - ``FILE``, is the name of the file that will be used for benchmarking. A file by this name should be in your benchmarking bucket in the ``uploads/`` folder.    

Reviewer: ```suggestion
    - ``FILE`` is the name of the file that will be used for benchmarking. A file by this name should be in your benchmarking bucket in the ``uploads/`` folder.    
```

    - ``COUNT`` specifies how many times the ``FILE`` will be submitted to the cluster for processing.

3. Deploy a DeepCell Kiosk as you normally would. While navigating the cluster configuration menu, pay special attention to two configuration settings:

    - The bucket name you provide should be that of the benchmarking bucket from step 1.
    - The Maximum Number of GPUs has a strong effect on benchmarking time, by effectively limiting how large the cluster can scale.

> 4. Once the cluster has finished deploying, drop to the ``Shell`` via the DeepCell Kiosk main menu and create the benchmarking pod to begin the benchmarking process by executing the following command: ``kubectl scale deployment benchmarking --replicas=1``.

Reviewer: ```suggestion
4. Once the cluster has deployed successfully, drop to the ``Shell`` via the DeepCell Kiosk main menu and begin the benchmarking process by executing the following command: ``kubectl scale deployment benchmarking --replicas=1``.
```

They don't really need to know that they are creating a benchmarking pod.

>     - The bucket name you provide should be that of the benchmarking bucket from step 1. To ensure the benchmarking process works properly, this bucket should not be attached to any other cluster you might have running.

Reviewer: Again, there is no risk in sharing buckets with other clusters.

```suggestion
    - The bucket name you provide should be that of the benchmarking bucket from step 1.
```

Author: Ok, that makes sense. As a response to both this comment and the one above it, do I still want to recommend that users create a new bucket?

Reviewer: I don't think its necessary to require they create one. Just tell them to either create a new bucket or use an existing one.

> 4. Once the cluster has finished deploying, drop to the ``Shell`` via the DeepCell Kiosk main menu and execute the following command, ``kubectl scale deployment benchmarking --replicas=1``, to create the benchmarking pod and begin the benchmarking process.

Reviewer: Something about the commas is off. Maybe reorder it somehow?

>  4. Once the cluster has finished deploying, drop to the ``Shell`` via the DeepCell Kiosk main menu and create the benchmarking pod to begin the benchmarking process by executing the following command: ``kubectl scale deployment benchmarking --replicas=1``.

> 5. Now, it's time to wait. Benchmarking jobs can take a day or more, depending on the conditions (# of images and max # of GPUs) chosen. To monitor the status of your benchmarking job, drop to the ``Shell`` within the DeepCell Kiosk main menu and execute the command ``stern benchmarking -s 10m``.  This will show you the most recent log output from the `benchmarking` pod. When benchmarking has finished, the final line in the log should be ``Uploaded [FILEPATH] to [BUCKET] in [SECONDS] seconds.``, where ``[FILEPATH]`` is the location in ``[BUCKET]`` where the benchmarking data has been saved.

Reviewer: ```suggestion
5. Benchmarking jobs can take a day or more, depending on the conditions (# of images and max # of GPUs) chosen. To monitor the status of your benchmarking job, drop to the ``Shell`` within the DeepCell Kiosk main menu and execute the command ``stern benchmarking -s 10m``. This will show you the most recent log output from the `benchmarking` pod. When benchmarking has finished, the final line in the log should be ``Uploaded [FILEPATH] to [BUCKET] in [SECONDS] seconds.``, where ``[FILEPATH]`` is the location in ``[BUCKET]`` where the benchmarking data has been saved.
```

6. Now that data has been generated for your benchmarking run and saved in your bucket, you can download and analyze it. Two top-level fields in this large JSON file that are probably of interest are:

    - The exact running time of the benchmarking procedure is given in seconds as the value of the ``time_elapsed`` field.

>     - A slight underestimate of the total costs of the benchmarking run can be found as the value to the `total_node_and_networking_costs` field. (Note that the total_node_and_networking_costs does not include Storage, Operation, or Storage Egress Fees. These extra fees `can be calculated <https://github.com/vanvalenlab/publication-figures/blob/383a90149eb86d4a0a697395edffb32d383bb1ca/figure_generation/data_extractor.py#L318>`_ after the fact by using the `Google Cloud guidelines <https://cloud.google.com/vpc/network-pricing#general>`_.)

Reviewer: ```suggestion
    - A slight underestimate of the total costs of the benchmarking run can be found as the value to the ``total_node_and_networking_costs`` field. (Note that the total_node_and_networking_costs does not include Storage, Operation, or Storage Egress Fees. These extra fees `can be calculated <https://github.com/vanvalenlab/publication-figures/blob/383a90149eb86d4a0a697395edffb32d383bb1ca/figure_generation/data_extractor.py#L318>`_ after the fact by using the `Google Cloud guidelines <https://cloud.google.com/vpc/network-pricing#general>`_.)
```

Reviewer: Would another bullet list work here?

> - `time_elapsed`: ...
> - `total_node_and_networking_costs`: ...

>     - A slight underestimate of the total costs of the benchmarking run can be found as the value to the `total_node_and_networking_costs` field. (Note that the total_node_and_networking_costs does not include Storage Fees, Operation Fees, or Storage Egress Fees. These `were calculated <https://github.com/vanvalenlab/publication-figures/blob/383a90149eb86d4a0a697395edffb32d383bb1ca/figure_generation/data_extractor.py#L318>`_ after the fact by using the `Google Cloud guidelines <https://cloud.google.com/vpc/network-pricing#general>`_.)

Reviewer 2: - The use of the pronoun "These" here is ambiguous (e.g. are we talking about the costs with or without the storage, operation, and egress fees?). Let's be explicit. 
- No need to use "Fees" 3 times in a list ("...Storage, Storage Egress, or Operation Fees.")
- "were calculated..." is another possible source of confusion. Should this be, "can be calculated..."?

5. any Persistent Disks associated with your cluster

While we hope this list is comprehensive, there could be some lingering resources used by Google Cloud and not deleted automatically that we're not aware of.

Benchmarking the DeepCell Kiosk
---------------------------------------------------------

> The DeepCell Kiosk comes with a utility for benchmarking the scalability and performance of a deep learning workflow. To reproduce the cost and timing benchmarks reported in the DeepCell Kiosk paper, please refer to our `figure creation repository <https://github.com/vanvalenlab/kiosk-console/pull/github.com/vanvalenlab/publication-figures>`_. To run your own benchmarking, please read below.

Reviewer 2: I believe this hyperlink is incorrect

Author: Good catch. That link was completely corrupted.

> The DeepCell Kiosk comes with a utility for benchmarking the scalability and performance of a deep learning workflow. To reproduce the cost and timing benchmarks reported in the DeepCell Kiosk paper, please refer to `the 2020-Bannon_et_al-Kiosk folder of our figure creation repository <https://github.com/vanvalenlab/publication-figures/2020-Bannon_et_al-Kiosk>`_. To run your own benchmarking, please read below.

Reviewer: ```suggestion
The DeepCell Kiosk comes with a utility for benchmarking the scalability and performance of a deep learning workflow. To reproduce the cost and timing benchmarks reported in the DeepCell Kiosk paper, please refer to `the 2020-Bannon_et_al-Kiosk folder of our figure creation repository <https://github.com/vanvalenlab/publication-figures/tree/master/2020-Bannon_et_al-Kiosk>`_. To run your own benchmarking, please read below.
```

1. If you don't already have a cloud storage bucket for use with the DeepCell Kiosk, you should create one now. It's fine to reuse this bucket across multiple DeepCell Kiosk clusters.

2. There are three variables in the benchmarking pod's YAML file, ``conf/helmfile.d/0410.benchmarking.yaml``, that may need to be customized before benchmarking:

>     - ``MODEL`` is the model name and version that will be used in benchmarking. The model you choose should be present in the ``models/`` folder of your benchmarking bucket. See the `Van Valen Lab's benchmarking bucket <https://console.cloud.google.com/storage/browser/kiosk-benchmarking>`_ for an example.

Reviewer: This URL doesn't provide access. I think there's another URL structure we can use that does allow access to unauthenticated accounts? The bucket has public read.

Reviewer: Nevermind, it was a bucket configuration issue. I made the bucket public read for all users and the link works.

Author: Yeah, good catch. It should be `http://storage.googleapis.com/kiosk-benchmarking`. That just returns a raw XML file, though. You think we need to provide some guidance on how to navigate that file?

Reviewer: I'm getting confused by where comments go, but I mentioned elsewhere that the issue was actually the access rights to the bucket. It's been resolved on the bucket side and the original link should work as intended.

    - ``FILE`` is the name of the file that will be used for benchmarking. A file by this name should be in your benchmarking bucket in the ``uploads/`` folder.    
    - ``COUNT`` specifies how many times the ``FILE`` will be submitted to the cluster for processing.

3. Deploy a DeepCell Kiosk as you normally would. While navigating the cluster configuration menu, pay special attention to two configuration settings:

    - The bucket name you provide should be that of the benchmarking bucket from step 1.

>     - The Maximum Number of GPUs has a strong effect on benchmarking time, by effectively limiting how large the cluster can scale.

Reviewer: Is the comma necessary?

> 4. Once the cluster has deployed successfully, drop to the ``Shell`` via the DeepCell Kiosk main menu and begin the benchmarking process by executing the following command: ``kubectl scale deployment benchmarking --replicas=1``.

Reviewer 2: no period here

Reviewer 2: ```suggestion
4. Once the cluster has deployed successfully, drop to the ``Shell`` via the DeepCell Kiosk main menu and begin the benchmarking process by executing the following command: ``kubectl scale deployment benchmarking --replicas=1``
```

> 5. Benchmarking jobs can take a day or more, depending on the conditions (# of images and max # of GPUs) chosen. To monitor the status of your benchmarking job, drop to the ``Shell`` within the DeepCell Kiosk main menu and execute the command ``stern benchmarking -s 10m``. This will show you the most recent log output from the `benchmarking` pod. When benchmarking has finished, the final line in the log should be ``Uploaded [FILEPATH] to [BUCKET] in [SECONDS] seconds.``, where ``[FILEPATH]`` is the location in ``[BUCKET]`` where the benchmarking data has been saved.

Reviewer 2: are we ok with being informal here? (# instead of number)

Author: Good call. I changed the symbols to words, to err on the side of professionalism.

Benchmarking the DeepCell Kiosk
---------------------------------------------------------

> The DeepCell Kiosk comes with a utility for benchmarking the scalability and performance of a deep learning workflow. To reproduce the cost and timing benchmarks reported in the DeepCell Kiosk paper, please refer to the ``2020-Bannon_et_al-Kiosk`` folder of our `figure creation repository <https://github.com/vanvalenlab/publication-figures>`_. To run your own benchmarking, please read below.

Reviewer: Could leave the "Please refer to" text the same, and just change the URL to be the subdirectory?

Reviewer 2: co-sign on that

2. There are three variables in the benchmarking pod's YAML file, ``conf/helmfile.d/0410.benchmarking.yaml``, that may need to be customized before benchmarking:

>     - ``MODEL`` is the model name and version that will be used in benchmarking. The model you choose should be present in the ``models/`` folder of your benchmarking bucket. See the `Van Valen Lab's benchmarking bucket <https://storage.googleapis.com/kiosk-benchmarking>`_ for an example.

Reviewer: ```suggestion
    - ``MODEL`` is the model name and version that will be used in benchmarking. The model you choose should be present in the ``models/`` folder of your benchmarking bucket. See the `Van Valen Lab's benchmarking bucket <https://console.cloud.google.com/storage/browser/kiosk-benchmarking>`_ for an example.
```

Actually, the last link worked, but the problem was on the bucket side. I had only given read access to particular files, so the link to the bucket itself did not work. I've fixed the read-access and now the former link should work.
